### ISSUE TITLE
Address notifications disconnected after one send.

### ISSUE THREAD

**Issue by evoskuil** [libbitcoin/libbitcoin-server#343]
Address subscription works correctly, and is actually just a specialized query against the query service. Notifications against subscriptions work properly for the fist notification. After that the zeromq pipe between the notification dealer and the query router becomes disconnected (globally, at least for a given receiver host), preventing the routing of subsequent notifications.

The break is not detected as an error due to the configuration of the router. A missing pipe results in a dropped message. This is intentional as otherwise the connection will block until the pipe is found (i.e. when using the zeromq mandatory router configuration setting). The pipe should presumably never be lost as it exists between the internal notification dealer on one thread and the internal query router on another thread. The dealer connects dynamically (i.e. on each send) to the router, as the dealer is invoked on any network thread, and that connection always succeeds.

The following section of zeromq code fails (silently) due to the `if (it != outpipes.end ()) ...` condition returning false. There is one pipe in the collection for both the first (successful) case and the later (failure) cases, even if the route and message is the same in all cases.
```cpp
int zmq::router_t::xsend (msg_t *msg_)
{
    //  If this is the first part of the message it's the ID of the
    //  peer to send the message to.
    if (!more_out) {
        zmq_assert (!current_out);

        //  If we have malformed message (prefix with no subsequent message)
        //  then just silently ignore it.
        //  TODO: The connections should be killed instead.
        if (msg_->flags () & msg_t::more) {

            more_out = true;

            //  Find the pipe associated with the identity stored in the prefix.
            //  If there's no such pipe just silently ignore the message, unless
            //  router_mandatory is set.
            blob_t identity ((unsigned char*) msg_->data (), msg_->size ());
            outpipes_t::iterator it = outpipes.find (identity);

            if (it != outpipes.end ()) {
                current_out = it->second.pipe;
                if (!current_out->check_write ()) {
                    it->second.active = false;
                    current_out = NULL;
                    if (mandatory) {
                        more_out = false;
                        errno = EAGAIN;
                        return -1;
                    }
                }
            }
            else
            if (mandatory) {
                more_out = false;
                errno = EHOSTUNREACH;
                return -1;
            }
        }

        int rc = msg_->close ();
        errno_assert (rc == 0);
        rc = msg_->init ();
        errno_assert (rc == 0);
        return 0;
    }

    //  Ignore the MORE flag for raw-sock or assert?
    if (options.raw_socket)
        msg_->reset_flags (msg_t::more);

    //  Check whether this is the last part of the message.
    more_out = msg_->flags () & msg_t::more ? true : false;

    //  Push the message into the pipe. If there's no out pipe, just drop it.
    if (current_out) {

        // Close the remote connection if user has asked to do so
        // by sending zero length message.
        // Pending messages in the pipe will be dropped (on receiving term- ack)
        if (raw_socket && msg_->size() == 0) {
            current_out->terminate (false);
            int rc = msg_->close ();
            errno_assert (rc == 0);
            rc = msg_->init ();
            errno_assert (rc == 0);
            current_out = NULL;
            return 0;
        }

        bool ok = current_out->write (msg_);
        if (unlikely (!ok)) {
            // Message failed to send - we must close it ourselves.
            int rc = msg_->close ();
            errno_assert (rc == 0);
            current_out = NULL;
        } else {
          if (!more_out) {
              current_out->flush ();
              current_out = NULL;
          }
        }
    }
    else {
        int rc = msg_->close ();
        errno_assert (rc == 0);
    }

    //  Detach the message from the data buffer.
    int rc = msg_->init ();
    errno_assert (rc == 0);

    return 0;
}
```

**Comment by evoskuil** [comment]
Hi @bluca Might I bother you for some insight into why the pipe would not be found in the above scenario?

**Comment by bluca** [comment]
Sorry, can't tell off the top of my head
Can you reproduce this problem with a small-ish test case?

**Comment by evoskuil** [comment]
@bluca No worries, I'll spend some time on this in a few days and see if I can provide a repo.

**Comment by evoskuil** [comment]
Tracking: https://github.com/zeromq/libzmq/issues/2443

**Comment by evoskuil** [comment]
I've been working through a full test suite for the authenticator and socket classes in libbitcoin-protocol. At this point I'm pretty sure that socket configuration and routing is incorrect for the intended asynchronous broker pattern.

**Comment by evoskuil** [comment]
This was a routing failure, resolved by changing router to dealer. Fixed in master for v3.1.
